### DISCUSSION TITLE
ws with extractor Query

### DISCUSSION THREAD

**Discussion by 799189288** [tokio-rs/axum#2061] [Q&A] [answered]
- [x] I have looked for existing issues (including closed) about this

### Version
axum = { version = "0.6", features = ["multipart", "ws", "headers"] }

### Platform
windows 11 64

### Crates
axum = { version = "0.6", features = ["multipart", "ws", "headers"] }

### Description
When I use the query extractor, the websocket link directly fails, but when I try to use Path, it succeeds.
query url like this: ws://localhost:3000/room?token=123
path url like this: ws://localhost:3000/room/123
![image](https://github.com/tokio-rs/axum/assets/73056742/60ce4593-e886-484c-b6e4-874effcc385d)
![image](https://github.com/tokio-rs/axum/assets/73056742/2c5fd0aa-3103-4cbb-8c6f-d0389dfa4be7)
![image](https://github.com/tokio-rs/axum/assets/73056742/2dbcf122-4874-4924-be65-7fa8cf312915)
![image](https://github.com/tokio-rs/axum/assets/73056742/7a3e41f5-0b06-459d-ab2c-8233a3237d99)

**Comment by jplatte** [comment]
The field in your `Token` struct is called `content`, but you're supplying `token=...`, not `content=...` in the query string. If you rename the field to match, or add `#[serde(rename = "token")]` to it, it should work.

**Reply by 799189288** [reply]
Thank you for your reply. I tried to change the `token` to `content` again, but it still didn't take effect. I also failed to use the `#[serde(rename = "token")]`
when i use `
pub struct Token {
    content: String,
}`
error: 
`WebSocket connection to 'ws://localhost:3000/room?content=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiI3OTkxODkyODgiLCJjb21wYW55IjoiNzk5MTg5Mjg4QHFxLmNvbSIsImV4cCI6MjAwMDAwMDAwMH0.TrA3q4dSu83JVTbCTqNbkTLJF7ebvnqY-RbjSoiEkH0' failed: `
when i use `
pub struct Token {
    #[serde(rename = "token")]
    content: String,
}
`
error: `WebSocket connection to 'ws://localhost:3000/room?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiI3OTkxODkyODgiLCJjb21wYW55IjoiNzk5MTg5Mjg4QHFxLmNvbSIsImV4cCI6MjAwMDAwMDAwMH0.TrA3q4dSu83JVTbCTqNbkTLJF7ebvnqY-RbjSoiEkH0' failed: `

**Comment by davidpdrsn** [comment, marked as answer]
Here is an example that does work https://gist.github.com/davidpdrsn/b8bbf57a2daacef5f58bfbfd42b6a2e3. Its based on https://github.com/tokio-rs/axum/blob/main/examples/websockets/src/main.rs

**Reply by 799189288** [reply]
Thank you. By the way, how can I directly add a token to the request header through the websocket link. And use TypedHeader to extract. Can this only be achieved through nginx? Is there any other method.

**Reply by davidpdrsn** [reply]
> By the way, how can I directly add a token to the request header through the websocket link

I don't know, haven't used javascript websockets much.
